File: Editorial/790_Domino_and_Tromino_Tiling/Python3/4_Matrix_Exponentiation.py
class Solution:
    def numTilings(self, n: int) -> int:
        MOD = 1_000_000_007
        SQ_MATRIX = [[1, 1, 2], [1, 0, 0], [0, 1, 1]] # Initialize square matrix
        SIZE = 3 # Width/Length of square matrix

        def matrix_product(m1, m2):  
            """Return product of 2 square matrices."""
            nonlocal MOD, SIZE
            # Result matrix `ans` will also be a square matrix with same dimensions.
            ans = [[0] * SIZE for _ in range(SIZE)]  
            for row in range(SIZE):
                for col in range(SIZE):
                    cur_sum = 0
                    for k in range(SIZE):
                        cur_sum += (m1[row][k] * m2[k][col]) % MOD
                    ans[row][col] = cur_sum
            return ans

        def matrix_expo(n):  
            """Perform matrix multiplication n times."""
            cur = SQ_MATRIX
            for _ in range(1, n):
                cur = matrix_product(cur, SQ_MATRIX)
            # The answer will be cur[0][0] * f(2) + cur[0][1] * f(1) + cur[0][2] * p(2)
            return (cur[0][0] * 2 + cur[0][1] * 1 + cur[0][2] * 1) % MOD

        # Handle base cases
        if n <= 2:
            return n  
        
        return matrix_expo(n - 2)

File: Editorial/790_Domino_and_Tromino_Tiling/CPP/4_Matrix_Exponentiation.cpp
class Solution {
public:
    // To avoid confusing point syntax, we use 2d-vector instead
    vector<vector<long>> SQ_MATRIX { // Initialize square matrix
        {1, 1, 2}, 
        {1, 0, 0}, 
        {0, 1, 1},
    };
    int MOD = 1'000'000'007;
    int SIZE = 3; // Width/Lenght of the square matrix
    
    // Return product of 2 square matrices.
    vector<vector<long>> matrixProduct(vector<vector<long>> m1, 
                                       vector<vector<long>> m2) {
        
        // Result matrix `ans` will also be a square matrix with same dimension
        vector<vector<long>> ans = SQ_MATRIX;  
        for (int row = 0; row < SIZE; ++row) {
            for (int col = 0; col < SIZE; ++col) {
                long curSum = 0;
                for (int k = 0; k < SIZE; ++k) {
                    curSum = (curSum + m1[row][k] * m2[k][col]) % MOD;
                }
                ans[row][col] = curSum;
            }
        }
        return ans;
    }
    
    // Return answer after n times matrix multiplication.
    int matrixExpo(int n) {  
        vector<vector<long>> cur = SQ_MATRIX;
        for (int i = 1; i < n; ++i) {
            cur = matrixProduct(cur, SQ_MATRIX);
        }
        // The answer will be cur[0][0] * f(2) + cur[0][1] * f(1) + cur[0][2] * p(2) 
        return static_cast<int>((cur[0][0] * 2 + cur[0][1] * 1 + cur[0][2] * 1) % MOD);  
    }
    
    int numTilings(int n) {
        // Handle base cases
        if (n <= 2) {  
            return n;
        } 
        return matrixExpo(n-2);
    }
};
